### PR TITLE
feat: Regen721Engine contract

### PR DESCRIPTION
**Pull request by thanhle-ancient8**

### CODE REVIEW

File: contracts/regen-721-engine/Regen721Engine.sol [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  modifier canSetBurnToken(address token) {
+    _checkRole(OPERATOR_ROLE, msg.sender);
+    require(
+      IERC165(token).supportsInterface(type(ERC721Burnable).interfaceId),
+      'Token is not burnable'
+    );
+    _;
+  }
+
+  modifier canBurn(uint256 amount, uint256[] calldata burnTokenIds) {
```
Comment (robertnguyen98) [inline contracts/regen-721-engine/Regen721Engine.sol:39]: why don't use burnTokenIds.length?

File: contracts/regen-721-engine/Regen721Engine.sol [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    uint256[] calldata burnTokenIds,
+    bytes32[] memory proof
+  )
+    external
+    canBurn(amount, burnTokenIds)
+    onlyWhitelist(seasonId, roundIdx, total, proof)
+  {
+    _payForClaim(seasonId, roundIdx, amount);
+    _burnTokens(burnTokens[seasonId], burnTokenIds);
+    _updateClaimedData(seasonId, roundIdx, total, amount);
```
Comment (robertnguyen98) [inline contracts/regen-721-engine/Regen721Engine.sol:112]: double-check for existing of _mint function here or not

File: contracts/extensions/interfaces/IDrop.sol [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+   * If the admin has incorect values while creating the season,
+   * E.g. Round 2 has less than or equal to round 1, tx should be reverted.
+   * @param startTime The start time of the round.
+   * @param endTime The end time of the round.
+   * @param pricePerToken The price for each drop of token.
+   * @param currency The currency to pay. Can be address(0) to use ETH.
+   * @param merkleRoot The merkle root of the round.
+   */
+  struct Round {
+    uint256 cummulativeMaxSupply;
```
Comment (robertnguyen98) [inline contracts/extensions/interfaces/IDrop.sol:29]: How do I know how many tokens is distributed in every round?

Comment (thanhle-ancient8) [reply contracts/extensions/interfaces/IDrop.sol:29]: we don't need to care about it, or just fetch the event of it round to count.

File: contracts/extensions/interfaces/IAssetManager.sol [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  /* ================ Events ================ */
+  event TokenWithdrawn(
+    address indexed recipient,
+    address indexed currency,
+    uint256 amount
+  );
+
+  event NftWithdrawn(
+    address indexed recipient,
+    address nftAddres,
```
Comment (robertnguyen98) [inline contracts/extensions/interfaces/IAssetManager.sol:14]:
```suggestion
    address nftAddress,
```

File: contracts/extensions/Drop.sol [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+  }
+
+  /**
+   * @param season The season data to create.
+   * The `totalDistributed` will be ignored.
+   * @return The ID of the created season.
+   */
+  function _createSeason(
+    Season memory season
+  ) internal canCreateSeason(season) returns (uint256) {
```
Comment (robertnguyen98) [inline contracts/extensions/Drop.sol:224]:
```suggestion
  ) internal canCreateSeason(season) returns (uint256 seasonId) {
```

File: contracts/extensions/Drop.sol [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    uint256 seasonId = nextSeasonId++;
```
Comment (robertnguyen98) [inline contracts/extensions/Drop.sol:225]:
```suggestion
    seasonId = nextSeasonId++;
```

File: contracts/extensions/Drop.sol [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+    seasons[seasonId] = Season({
+      maxSupply: season.maxSupply,
+      totalDistributed: 0
+    });
+    emit SeasonCreated(seasonId, season.maxSupply);
+    return seasonId;
```
Comment (robertnguyen98) [inline contracts/extensions/Drop.sol:231]:
```suggestion
```

File: contracts/extensions/Drop.sol [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+  modifier canClaim(
+    uint256 seasonId,
+    uint256 roundIdx,
+    uint256 total,
+    uint256 amount
+  ) virtual {
+    _checkSeasonExists(seasonId);
+    _checkActiveRound(seasonId, roundIdx);
+    _checkValidClaim(seasonId, roundIdx, total, amount);
```
Comment (robertnguyen98) [inline contracts/extensions/Drop.sol:68]: It's unusual here (Define functions to call in modifiers) . Modifier exist for running checkings before and after functions. I think these function should be made as the separated modifiers

File: contracts/regen-721-engine/Regen721Engine.sol [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+  /**
+   * @dev Set the burn token for a season.
+   * @param seasonId The ID of the season.
+   * @param setting The burn mint setting.
+   */
+  function setBurnMintSetting(
+    uint256 seasonId,
+    BurnMintSetting calldata setting
+  ) external canSetBurnMintSetting(setting) {
```
Comment (robertnguyen98) [inline contracts/regen-721-engine/Regen721Engine.sol:101]: should only allow set BurnMintSetting when season haven't started yet

File: contracts/regen-721-engine/Regen721Engine.sol [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  function createSeason(
+    Season calldata season,
+    BurnMintSetting calldata setting
+  ) external returns (uint256) {
+    burnMintSettings[nextSeasonId] = setting;
+    return _createSeason(season);
+  }
+
+  /**
+   * @dev Set the burn token for a season.
```
Comment (robertnguyen98) [inline contracts/regen-721-engine/Regen721Engine.sol:94]:
```suggestion
   * @dev Set the burn and mint tokens for a season.
```

File: contracts/regen-721-engine/Regen721Engine.sol [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    _burnTokens(setting.burnToken, burnTokenIds);
+    _updateClaimedData(seasonId, roundIdx, total, amount);
+    _mintTokens(setting.mintToken, amount);
+    emit TokensBurnedAndMinted(
+      seasonId,
+      setting.burnToken,
+      setting.mintToken,
+      amount
+    );
+    emit ItemsClaimed(msg.sender, seasonId, roundIdx, amount);
```
Comment (robertnguyen98) [inline contracts/regen-721-engine/Regen721Engine.sol:136]: I think 1 function should only emit 1 event. Consider ability to change 2 events into 1 event

File: contracts/extensions/Drop.sol [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+  }
+
+  /* ================ Admin Functions ================ */
+  /**
+   * @dev External function for `_createSeason`.
+   */
+  function createSeason(
+    Season calldata season
+  ) external virtual returns (uint256) {
+    return _createSeason(season);
```
Comment (robertnguyen98) [inline contracts/extensions/Drop.sol:105]: Why do we need to implement an external and internal function like this?

File: contracts/extensions/Drop.sol [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+   */
+  function _setRounds(
+    uint256 seasonId,
+    Round[] memory rounds
+  ) internal canSetRounds(seasonId, rounds) {
+    for (uint i = 0; i < rounds.length; i++) {
+      seasonRounds[seasonId].push(rounds[i]);
+    }
+    emit RoundsSet(seasonId);
+  }
```
Comment (robertnguyen98) [inline contracts/extensions/Drop.sol:259]: Be aware that _setRound will make data changed, and we can get the exact round info when use roundIdx from ItemsClaimed event

File: contracts/extensions/Drop.sol (hunk 8 above)
Comment (thanhle-ancient8) [reply contracts/extensions/Drop.sol:68]: These check functions are fine and they reduce the call stack, I used to use modifiers but got errors.

File: contracts/regen-721-engine/Regen721Engine.sol (hunk 9 above)
Comment (thanhle-ancient8) [reply contracts/regen-721-engine/Regen721Engine.sol:101]: No it's anytime, let me check.

File: contracts/regen-721-engine/Regen721Engine.sol (hunk 11 above)
Comment (thanhle-ancient8) [reply contracts/regen-721-engine/Regen721Engine.sol:136]: `ItemsClaimed` from Drop and `TokensBurnedAndMinted` from child contract, what do you think?

File: contracts/extensions/Drop.sol (hunk 12 above)
Comment (thanhle-ancient8) [reply contracts/extensions/Drop.sol:105]: For reusing these logic.

File: contracts/extensions/Drop.sol (hunk 13 above)
Comment (thanhle-ancient8) [reply contracts/extensions/Drop.sol:259]: This is a bug, will fix it

File: contracts/extensions/Drop.sol [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+    uint256 seasonId,
+    Round[] memory rounds
+  ) internal view {
+    Season memory season = seasons[seasonId];
+    for (uint i = 0; i < rounds.length; i++) {
+      Round memory round = rounds[i];
+      /// Round data validations.
+      require(
+        round.cummulativeMaxSupply > 0,
+        'Round_checkActiveRoundtive max supply must be greater than 0'
```
Comment (robertnguyen98) [inline contracts/extensions/Drop.sol:157]: check wording here

File: contracts/extensions/Drop.sol [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+      );
+      require(
+        round.cummulativeMaxSupply <= season.maxSupply,
+        'Round: Exceeds season max supply'
+      );
+      /// Round relative data validations.
+      if (i > 0) {
+        require(
+          round.cummulativeMaxSupply > rounds[i - 1].cummulativeMaxSupply,
+          'Round_checkActiveRoundtive max supply must be greater than the previous round'
```
Comment (robertnguyen98) [inline contracts/extensions/Drop.sol:171]: check wording here

**Review by robertnguyen98 — APPROVED** [review]

**Pull request merged**